# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [12]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [42]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [43]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [44]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [45]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [37]:
import pandas as pd
import numpy as np
import scipy.stats as stats
poke = pd.read_csv('pokemon.csv')

In [27]:
poke

Name  Num   Type 1  Type 2  Total   HP  Attack  Defense  SpAtk  \
0     Bulbasaur    1    Grass  Poison    318   45      49       49     65   
1       Ivysaur    2    Grass  Poison    405   60      62       63     80   
2      Venusaur    3    Grass  Poison    525   80      82       83    100   
3    Charmander    4     Fire     NaN    309   39      52       43     60   
4    Charmeleon    5     Fire     NaN    405   58      64       58     80   
..          ...  ...      ...     ...    ...  ...     ...      ...    ...   
146     Dratini  147   Dragon     NaN    300   41      64       45     50   
147   Dragonair  148   Dragon     NaN    420   61      84       65     70   
148   Dragonite  149   Dragon  Flying    600   91     134       95    100   
149      Mewtwo  150  Psychic     NaN    680  106     110       90    154   
150         Mew  151  Psychic     NaN    600  100     100      100    100   

     SpDef  Speed  Stage  Legendary  
0       65     45      1      False  
1       80     60      2      False  
2      100     80      3      False  
3       50     65      1      False  
4       65     80      2      False  
..     ...    ...    ...        ...  
146     50     50      1      False  
147     70     70      2      False  
148    100     80      3      False  
149     90    130      1       True  
150    100    100      1      False  

[151 rows x 13 columns]

In [30]:
poke.head(5)

Name  Num Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0   Bulbasaur    1  Grass  Poison    318  45      49       49     65     65   
1     Ivysaur    2  Grass  Poison    405  60      62       63     80     80   
2    Venusaur    3  Grass  Poison    525  80      82       83    100    100   
3  Charmander    4   Fire     NaN    309  39      52       43     60     50   
4  Charmeleon    5   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [31]:
poke['Type 1'].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [32]:
poke['Type 1'].value_counts(normalize =True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [19]:
pd.crosstab(poke["Stage"], poke["Legendary"], margins=False, normalize=False)

Legendary  False  True
Stage                 
1             75     4
2             56     0
3             16     0

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [46]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [47]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [48]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [49]:
poke['Attack'].mean()

72.54966887417218

In [50]:
poke['Defense'].median()

65.0

In [51]:
poke['Speed'].mode()

0    70
1    90
dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [52]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [73]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [21]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [25]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

In [55]:
df['Hours Worked'].quantile(.50)

34.0

In [56]:
df['Hours Worked'].quantile(.75)

41.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [76]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [77]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [39]:
hp_range = poke['HP'].max()- poke['HP'].min()

print(hp_range)

240


In [32]:
poke['SpAtk'].mean()

67.13907284768212

In [33]:
poke['SpAtk'].std()

28.53419930191353

In [34]:
np.percentile(poke['Attack'], 80)

95.0

In [ ]:
#80% of the values in Attack are below 95.0

In [35]:
np.percentile(poke['Attack'], 15)

45.0

In [ ]:
#15 % of the values in Attack are below 45.0

In [41]:
poke['SpDef'].quantile(.50)

65.0

In [42]:
poke['SpDef'].median()

65.0

In [ ]:
# the second quartile and the median have the same value

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [43]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [44]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [45]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [81]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [82]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [83]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [84]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [85]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [86]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [87]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [88]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [89]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [90]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [38]:
import pandas as pd
import numpy as np
import scipy.stats as stats
df=pd.read_excel('babies.xlsx')
df

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
3     123        NaN       0  36.0    69.0   190.0    0.0
4     108      282.0       0  23.0    67.0   125.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 7 columns]

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [210]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

In [211]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


In [212]:
# the numerical variables make up the dataset

In [213]:
df.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [214]:
# Yes, there is  a missing data, the above numbers indicate the missing data.

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [215]:
# the shape of the dataset before missing value dropped

df.shape

(1236, 7)

In [216]:
df.dropna( inplace = True)

In [217]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
4  108      282.0       0  23.0    67.0   125.0    1.0
5  136      286.0       0  25.0    62.0    93.0    0.0

In [218]:
# the shape of the dataset after missing value dropped

df.shape


(1174, 7)

In [219]:
print(1236-1174)

62


In [220]:
# 62 rows dropped from the dataset

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [113]:
# the two qualitative variables are parity and smoke 
# these values can't be measured in numeric scale

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [132]:
df['parity'].replace([0, 1], ['N', 'Y'], inplace = True)

In [180]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [254]:
dfq= df.copy()

In [255]:
dfq

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
3     123        NaN       0  36.0    69.0   190.0    0.0
4     108      282.0       0  23.0    67.0   125.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 7 columns]

In [224]:
dfq.shape

(1236, 7)

In [225]:
q1 = dfq['bwt'].quantile(.25)
q3 = dfq['bwt'].quantile(.75)

print('Q1:', q1)
print('Q3:', q3)

Q1: 108.75
Q3: 131.0


In [226]:
iqr = q3- q1
print('IQR:', iqr)

IQR: 22.25


In [227]:
top = q3 + (iqr * 1.5)
bottom = q1- (iqr * 1.5)

print('Upper limit:', top)
print('Lower Limit:', bottom)

Upper limit: 164.375
Lower Limit: 75.375


In [236]:
# to determine outliers

iqr_outliers = dfq.loc[(dfq['bwt'] > top) | (dfq['bwt'] < bottom)].index

print('Index Value:', iqr_outliers)

Index Value: Int64Index([  59,  192,  239,  309,  361,  462,  500,  529,  556,  594,  632,
             709,  738,  739,  747,  814,  829,  904,  912,  978, 1020, 1021,
            1035, 1063, 1065, 1099, 1139, 1148, 1149, 1169, 1211],
           dtype='int64')


In [237]:
dfq.iloc[[ 192,  239,  309,  361,  462,  500,  529,  556,  594,  632,
             709,  738,  739,  747,  814,  829,  904,  912,  978, 1020, 1021,
            1035, 1063, 1065, 1099, 1139, 1148, 1149, 1169, 1211]]
 

bwt  gestation  parity   age  height  weight  smoke
192    75      239.0       0  26.0    63.0   124.0    1.0
239   173      293.0       0  30.0    63.0   110.0    0.0
309    71      281.0       0  32.0    60.0   117.0    1.0
361    71      234.0       0  32.0    64.0   110.0    1.0
462    68      223.0       0  32.0    66.0   149.0    1.0
500    69      232.0       0  31.0    59.0   103.0    1.0
529    71      277.0       0  40.0    69.0   135.0    0.0
556   174      281.0       0  37.0    67.0   155.0    0.0
594   170      303.0       1  21.0    64.0   129.0    0.0
632   176      293.0       1  19.0    68.0   180.0    0.0
709   166      299.0       0  26.0    68.0   140.0    0.0
738   167      288.0       1  19.0    63.0   117.0    0.0
739    71        NaN       0  19.0    64.0   120.0    0.0
747   174      288.0       0  25.0    61.0   182.0    0.0
814   165      282.0       0  29.0    66.0   145.0    0.0
829    62      228.0       0  24.0    61.0   107.0    0.0
904    72      271.0       0  39.0    61.0   136.0    0.0
912    58      245.0       0  34.0    64.0   156.0    1.0
978    55      204.0       0  35.0    65.0   140.0    0.0
1020   75      247.0       0  36.0    64.0   120.0    1.0
1021  169      296.0       0  33.0    67.0   185.0    0.0
1035   65      237.0       0  31.0    67.0   130.0    0.0
1063   73      277.0       0  29.0    65.0   145.0    0.0
1065   65      232.0       0  24.0    66.0   125.0    1.0
1099  174      284.0       0  39.0    65.0   163.0    0.0
1139   63      236.0       1  24.0    58.0    99.0    0.0
1148   72      266.0       1  25.0    66.0   200.0    1.0
1149   75      266.0       0  37.0    61.0   113.0    1.0
1169   71      254.0       0  19.0    61.0   145.0    1.0
1211   75      265.0       0  21.0    65.0   103.0    1.0

In [256]:
# drop outliers index for 
dfq = dfq.drop(iqr_outliers)

print(dfq.shape)


(1205, 7)


In [326]:
#finding ouliers for gestation 
dfq = df.copy()
dfq

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
3     123        NaN       0  36.0    69.0   190.0    0.0
4     108      282.0       0  23.0    67.0   125.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 7 columns]

In [327]:
dfq.shape

(1236, 7)

In [328]:
q1 = dfq['gestation'].quantile(.25)
q3 = dfq['gestation'].quantile(.75)
print('Q1:', q1)
print('Q3:', q3)


Q1: 272.0
Q3: 288.0


In [329]:
iqr =q3-q1
print(' IQR:', iqr)

 IQR: 16.0


In [330]:
top = q3 +(iqr * 1.5)
bottomm = q1- (iqr * 1.5)

print('Upper Limit:', top)
print('Lower Limit:', bottomm)

Upper Limit: 312.0
Lower Limit: 248.0


In [331]:
iqr_outliers = dfq.loc[(dfq['gestation'] > top) | (dfq['gestation'] < bottom)].index

print('Index Value:', iqr_outliers)

Index Value: Int64Index([  10,  152,  155,  198,  234,  240,  253,  279,  373,  394,  439,
             460,  685,  746,  761,  769,  778,  952,  969, 1002, 1026, 1044,
            1074, 1140, 1172, 1199, 1217, 1219],
           dtype='int64')


In [332]:
dfq.iloc[[ 59,  192,  309,  361,  462,  500,  529,  739,  829,  904,  912,
             978, 1020, 1035, 1063, 1065, 1139, 1148, 1149, 1169, 1211]]

bwt  gestation  parity   age  height  weight  smoke
59     75      232.0       0  33.0    61.0   110.0    0.0
192    75      239.0       0  26.0    63.0   124.0    1.0
309    71      281.0       0  32.0    60.0   117.0    1.0
361    71      234.0       0  32.0    64.0   110.0    1.0
462    68      223.0       0  32.0    66.0   149.0    1.0
500    69      232.0       0  31.0    59.0   103.0    1.0
529    71      277.0       0  40.0    69.0   135.0    0.0
739    71        NaN       0  19.0    64.0   120.0    0.0
829    62      228.0       0  24.0    61.0   107.0    0.0
904    72      271.0       0  39.0    61.0   136.0    0.0
912    58      245.0       0  34.0    64.0   156.0    1.0
978    55      204.0       0  35.0    65.0   140.0    0.0
1020   75      247.0       0  36.0    64.0   120.0    1.0
1035   65      237.0       0  31.0    67.0   130.0    0.0
1063   73      277.0       0  29.0    65.0   145.0    0.0
1065   65      232.0       0  24.0    66.0   125.0    1.0
1139   63      236.0       1  24.0    58.0    99.0    0.0
1148   72      266.0       1  25.0    66.0   200.0    1.0
1149   75      266.0       0  37.0    61.0   113.0    1.0
1169   71      254.0       0  19.0    61.0   145.0    1.0
1211   75      265.0       0  21.0    65.0   103.0    1.0

In [333]:
dfq = dfq.drop(iqr_outliers)

print(dfq.shape)

(1208, 7)


In [30]:
#finding ouliers in age

dfq = df.copy()
dfq. shape

(1236, 7)

In [31]:
q1 = dfq['age'].quantile(.25)
q3 = dfq['age'].quantile(.75)
print('Q1:', q1)
print('Q3:', q3)


Q1: 23.0
Q3: 31.0


In [32]:
iqr = q3 - q1
print('IQR:', iqr)

IQR: 8.0


In [33]:
top = q3 +(iqr * 1.5)
bottom = q1- (iqr * 1.5)

print('Upper Limit:', top)
print('Lower Limit:', bottom)

Upper Limit: 43.0
Lower Limit: 11.0


In [34]:
iqr_outliers = dfq.loc[(dfq['age'] > top) | (dfq['age'] < bottom)].index

print('Index Value:', iqr_outliers)

Index Value: Int64Index([633, 1070], dtype='int64')


In [35]:
dfq = dfq.iloc[[633,1070]]

In [36]:
dfq

bwt  gestation  parity   age  height  weight  smoke
633   143      294.0       0  44.0    65.0   145.0    0.0
1070  122      280.0       1  45.0    62.0   128.0    0.0

In [21]:
#
print(dfq.shape)

(1236, 7)


In [39]:
#finding outliers for height

dfq = df.copy()
dfq.shape

(1236, 7)

In [40]:
q1 = dfq['height'].quantile(.25)
q3 = dfq['height'].quantile(.75)
print('Q1:', q1)
print('Q3:', q3)

Q1: 62.0
Q3: 66.0


In [41]:
iqr = q3 - q1
print('IQR:', iqr)

IQR: 4.0


In [42]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)

print('Upper Limit:', top)
print('Lower Limit:', bottom)

Upper Limit: 72.0
Lower Limit: 56.0


In [43]:
iqr_outliers = dfq.loc[(dfq['height'] > top) | (dfq['height'] < bottom)].index

print('Index Value:', iqr_outliers)

Index Value: Int64Index([434, 1208], dtype='int64')


In [44]:
dfq = dfq.iloc[[434,1208]]

In [45]:
dfq

bwt  gestation  parity   age  height  weight  smoke
434   146      263.0       0  39.0    53.0   110.0    1.0
1208  141      281.0       0  29.0    54.0   156.0    1.0

In [46]:
dfq = dfq.drop(iqr_outliers)

In [387]:
#finding outliers in weight

dfq = df.copy()
dfq.shape

(1236, 7)

In [388]:
q1 = dfq['weight'].quantile(.25)
q3 = dfq['weight'].quantile(.75)

print('Q1:', q1)
print('Q3:', q3)

Q1: 114.0
Q3: 140.0


In [389]:
iqr = q3-q1
print('IQR:', iqr)

IQR: 26.0


In [390]:
top = q3 + (q3 * 1.5)
bottom = q1 - (q1 * 1.5)

print('Upper Limit:', top)
print('Lower Lilmit:', bottom)

Upper Limit: 350.0
Lower Lilmit: -57.0


In [392]:
iqr_outliers = dfq.loc[(dfq['weight'] > top) | (dfq['weight'] < bottom)].index
print('Index Value:',iqr_outliers )

Index Value: Int64Index([14, 125, 314, 344, 458, 595, 911], dtype='int64')


In [393]:
dfq.iloc[[14, 125, 314, 344, 458, 595, 911]]

bwt  gestation  parity   age  height  weight  smoke
14   114      273.0       0  30.0    63.0   498.0    0.0
125  115      283.0       0  25.0    61.0  1500.0    1.0
314  100      277.0       0  31.0    62.0   355.0    1.0
344  105      276.0       0  22.0    67.0   845.0    0.0
458   85      258.0       0  41.0    67.0   399.0    0.0
595  104      270.0       0  25.0    61.0   511.0    0.0
911  138      280.0       1  30.0    65.0   478.0    0.0

In [394]:
dfq = dfq.drop(iqr_outliers)

dfq.shape

(1229, 7)

7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [395]:
dfq['parity'].value_counts()

0    915
1    314
Name: parity, dtype: int64

In [396]:
dfq['smoke'].value_counts()

0.0    737
1.0    482
Name: smoke, dtype: int64

In [ ]:
# parity has highest frequency

In [397]:
dfq['parity'].value_counts(normalize =  True)

0    0.744508
1    0.255492
Name: parity, dtype: float64

In [398]:
dfq['smoke'].value_counts(normalize = True)

0.0    0.604594
1.0    0.395406
Name: smoke, dtype: float64

In [399]:
pd.crosstab(dfq['parity'], dfq['smoke'], margins = False, normalize = False)

smoke   0.0  1.0
parity          
0       544  361
1       193  121

In [ ]:
# both group have equal relative frequency
# 121 were pregnant before

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [ ]:
# Measures of Central Tendency are estimates of what a "typical" value/data point is within our set of data.
#A measure of central tendency is a number around which an entire sample of data is spread. 


In [400]:
dfq['bwt'].mean()

119.63873067534581

In [401]:
dfq['bwt'].median()

120.0

In [402]:
dfq['bwt'].mode()

0    115
1    117
dtype: int64

In [ ]:
#The median  is greater than the mean  it is a left-skewed data. 
#the median will be the middle value
#the mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency.
# 115 and 117 are the most frequently appear values in the data